docs/enhancements/0002_day0-integration.md

```diff
@@ -0,0 +1,257 @@
+systemd files for helping customers to deploy their day-1 kmod.
+
+What KMM doesn't offer today to customers is the deployment and management of
+day-0 kmods.
+
+Day-0 kmod are complex as they need to be present in the system at boot time,
+even before OCP is fully installed and certainly before KMM is deployed to the
+cluster, therefore, there is not much that KMM can do to help deploy those kmods but
+it can manage their lifecycle once a customer has installed a cluster with his
+day-0 kmods in it - this is what this enhancement is about.
+
+### Day-0 cluster background
+
+A customer wishing to deploy OCP with custom nodes will need to generate 2
+artifacts using [image-builder](https://www.redhat.com/en/topics/linux/what-is-an-image-builder)
+  * A custom ISO for the nodes
+  * A container image encapsulating the ostree-commit that is deployed in the ISO
+
+The ISO will be used by [assisted-installer](https://docs.openshift.com/container-platform/4.14/installing/installing_on_prem_assisted/installing-on-prem-assisted.html)
+to install the cluster and the container image will be used to notify
+[MCO](https://docs.openshift.com/container-platform/4.14/post_installation_configuration/machine-configuration-tasks.html#understanding-the-machine-config-operator)
+that the image on the nodes is a custom image.
+
+When a customers has installed a "day-0 cluster" as mentioned above, we can
+assume that there is a `MachineConfig` in the cluster that looks something like
+```
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: worker
+  name: 99-ybettan-external-image
+spec:
+  osImageURL: quay.io/ybettan/rhcos:414.92.202312311229-0
+```
+The important piece in it is `osImageURL: quay.io/ybettan/rhcos:414.92.202312311229-0`
+which tells MCO that the ostree-commit that is supposed to be running on
+the node is encapsulated in `quay.io/ybettan/rhcos:414.92.202312311229-0`.
+
+If MCO finds out that the OS on the node is different than the image specified
+in `osImageURL` then it will override the node OS based on the `MachineConfig`
+and reboot - we are going to leverage this mechanism for the integration with
+KMM.
+
+In a day-0-cluster scenario, the user has applied this `MachineConfig` to tell
+MCO he has modified the OS on the nodes and that it shoudn't override it with
+the RHCOS image from the payload.
+
+### Setting up the integration
+
+A user that wishes to hand over the management of his day-0 kmod to KMM will
+first need to make sure the `MachineConfig` can be reconize by KMM, therefore,
+he would add a `machineconfiguration.openshift.io/kmm-managed: "true"` label to
+it.
+
+KMM will watch for `MachineConfig`s in the cluster containing the
+`machineconfiguration.openshift.io/kmm-managed: "true"` label and create a new
+`ModuleDay0` CR in the cluster for each one of them.
+We will have a different controller for this new CRD in KMM.
```

Comment:
I wonder if we can inject all those manifests during the installation, similar to what happens with the /var partition creation (lets understand how it works - because this is something that happens day0 vs day2):
https://docs.openshift.com/container-platform/4.12/installing/installing_bare_metal/installing-bare-metal.html#installation-user-infra-machines-advanced_vardisk_installing-bare-metal

Reply:
What "all manifests"? there is only the `MachineConfig` which is infact injected to the assisted-cluster pre-installation and the use can set the `machineconfiguration.openshift.io/kmm-managed: "true"` label pre-installation as well if he wishes. It won't have any effect until KMM is installed in the cluster though.

Reply:
the question is whether that MC will become part of the ignition or only day2 (which affects number of reboots)

Reply:
I don't see any reference to the `MachineConfig` manifest in the discovery ignition. I am not sure when assisted applies the manifests that were added to the cluster pre-installation.

If that's answer your questions, that MC exists in the cluster as long as the cluster is customized and each change to it may trigger a reboot.